git dependencies does not respect dangerouslyAllowAllBuilds

### Verify latest release

- [x] I verified that the issue exists in the latest pnpm release

### pnpm version

_No response_

### Which area(s) of pnpm are affected? (leave empty if unsure)

_No response_

### Link to the code that reproduces this issue or a replay of the bug

_No response_

### Reproduction steps

1. Create a project.
2. Create `pnpm-workspace.yaml`:
```
dangerouslyAllowAllBuilds: true
```
3. Add a git dependency which have `prepare` script set. `$ pnpm add github:...`
4. `$ pnpm install` will produce an error:
```
ERR_PNPM_GIT_DEP_PREPARE_NOT_ALLOWED  Failed to prepare git-hosted package fetched from "git@github.com:...": The git-hosted package "..." needs to execute build scripts but is not in the "onlyBuiltDependencies" allowlist.
```

### Describe the Bug

- Git dependency with prepare script cannot be installed if `dangerouslyAllowAllBuilds` is true.
- There is no way to allow such packages to be installed when `dangerouslyAllowAllBuilds` is true.

While the error message suggests adding `onlyBuiltDependencies`, it will not work:
```
ERR_PNPM_CONFIG_CONFLICT_BUILT_DEPENDENCIES  Cannot have both neverBuiltDependencies and onlyBuiltDependencies
```

For reference, this is `pnpm-workspace.yaml` that I'm testing with:
```yaml
dangerouslyAllowAllBuilds: true

#onlyBuiltDependencies:
#  - my-packages
```

I think #10288 is related.

### Expected Behavior

- Git dependency should be able to run prepare script if `dangerouslyAllowAllBuilds` is set.
- As a workaround, some other option to allow git dependency to run prepare script that works alongside `dangerouslyAllowAllBuilds` would be ok.

### Which Node.js version are you using?

v24.12.0

### Which operating systems have you used?

- [ ] macOS
- [x] Windows
- [ ] Linux

### If your OS is a Linux based, which one it is? (Include the version if relevant)

_No response_

--- a/.changeset/early-numbers-look.md
+++ b/.changeset/early-numbers-look.md
@@ -1,0 +1,6 @@
+---
+"@pnpm/builder.policy": minor
+"pnpm": patch
+---
+
+Git dependencies with build scripts should respect the `dangerouslyAllowAllBuilds` settings [#10376](https://github.com/pnpm/pnpm/issues/10376).

--- a/builder/policy/src/index.ts
+++ b/builder/policy/src/index.ts
@@ -4,11 +4,13 @@ import fs from 'fs'
 
 export function createAllowBuildFunction (
   opts: {
+    dangerouslyAllowAllBuilds?: boolean
     neverBuiltDependencies?: string[]
     onlyBuiltDependencies?: string[]
     onlyBuiltDependenciesFile?: string
   }
 ): undefined | AllowBuild {
+  if (opts.dangerouslyAllowAllBuilds) return () => true
   if (opts.onlyBuiltDependenciesFile != null || opts.onlyBuiltDependencies != null) {
     const onlyBuiltDeps = opts.onlyBuiltDependencies ?? []
     if (opts.onlyBuiltDependenciesFile) {

--- a/builder/policy/test/index.ts
+++ b/builder/policy/test/index.ts
@@ -55,3 +55,11 @@ it('should onlyBuiltDependencies set via a file and config', () => {
 it('should return undefined if no policy is set', () => {
   expect(createAllowBuildFunction({})).toBeUndefined()
 })
+
+it('should allow everything when dangerouslyAllowAllBuilds is true', () => {
+  const allowBuild = createAllowBuildFunction({
+    dangerouslyAllowAllBuilds: true,
+  })
+  expect(typeof allowBuild).toBe('function')
+  expect(allowBuild!('foo', '1.0.0')).toBeTruthy()
+})

--- a/pnpm/test/install/lifecycleScripts.ts
+++ b/pnpm/test/install/lifecycleScripts.ts
@@ -299,3 +299,14 @@ test('the list of ignored builds is preserved after a repeat install', async () 
     'esbuild@0.25.0',
   ])
 })
+
+test('git dependencies with preparation scripts should be installed when dangerouslyAllowAllBuilds is true', async () => {
+  prepare({})
+  await writeYamlFile('pnpm-workspace.yaml', { dangerouslyAllowAllBuilds: true })
+
+  // 'test-git-fetch' has a prepare script that builds the package.
+  const result = execPnpmSync(['add', 'https://github.com/pnpm/test-git-fetch.git#8b333f12d5357f4f25a654c305c826294cb073bf'])
+
+  expect(result.status).toBe(0)
+  expect(fs.existsSync('node_modules/test-git-fetch/dist/index.js')).toBeTruthy()
+})

